Ksp 190 host should be able to cancel an event

ivancuadra869 opened the pull request:
I left in some front-end functionality for the sake of testing. Also because the front-end is changing to tailwind.

ivancuadra869 committed:
Add delete event-plan functionality

ivancuadra869 committed:
Add delete event functionality

jesc0bar commented:
Finalized event was not deleted on my end.

navarrojandg commented on apps/api/src/modules/etl/event-plans/create.ts:
```diff
@@ -90,7 +90,7 @@ export const etlEventPlansCreate = async (
         ...restOfParams,
         hostId,
-        inviteesByUserId,
+        invitees: inviteesByUserId,
```
Because there are a handful of times when we deal with invitees, such has sending out emails or transforming them from invitees to guests, its better to be explicit about the type even in the variable name, so that when we read this document its very clear that these invitees are represented by their user ids

navarrojandg commented on apps/api/src/modules/etl/event-plans/delete.ts:
```diff
@@ -0,0 +1,67 @@
+
+type Params = {
+  eventPlanId: EventPlanId;
+  invitees: UserId[];
```
Same note here -> inviteesByUserId is arguably a better name

Better to lean on best practices over language features. For example better to give a better variable name than to rely on Typescript's type features. We should use Typescript's type to help support our code rather than the other way around.

navarrojandg replied:
Sorry just realizing it now but I don't even think we need to pass in an array of invitees.

We really only need to pass in the eventPlanId and then maybe the userId (userId of whoever is making the call so we can hackily check if they are the host or not).

Use a transaction if you can to fetch the event-plan doc corresponding to the params.eventPlanId and thats where we get the inviteesByUserId from.

ivancuadra869 commented:
Okay, I'll make the changes. I'll have to change the name in the types in
the interface folder

On Sat, Apr 9, 2022, 8:53 AM Jan Navarro ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In apps/api/src/modules/etl/event-plans/delete.ts
> <https://github.com/CSUN-COMP-490-Team-KBBQ/wya/pull/10#discussion_r846650835>
> :
>
> > @@ -0,0 +1,67 @@
> +import assert from 'assert';
> +import Debug from 'debug';
> +import { App } from 'firebase-admin/app';
> +import { getAuth as getFirebaseAuth } from 'firebase-admin/auth';
> +import { getFirestore as getFirebaseFirestore } from 'firebase-admin/firestore';
> +
> +import { EventPlanId, UserId } from '../../../interfaces';
> +import { ApiError, makeApiError } from '../../../../lib/errors';
> +
> +type Params = {
> +  eventPlanId: EventPlanId;
> +  invitees: UserId[];
>
> Same note here -> inviteesByUserId is arguably a better name
>
> Better to lean on best practices over language features. For example
> better to give a better variable name than to rely on Typescript's type
> features. We should use Typescript's type to help support our code rather
> than the other way around.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CSUN-COMP-490-Team-KBBQ/wya/pull/10#pullrequestreview-937153730>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASZ432DDPCHRO6JD5GMW6BDVEGRWNANCNFSM5S6MO4GA>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

navarrojandg commented on apps/api/src/modules/etl/event-plans/delete.ts:
```diff
@@ -0,0 +1,67 @@
+    if (invitees !== undefined) {
+      inviteesAndHost.push(...invitees);
+    }
+    for (const uid of inviteesAndHost) {
```
`for (const uid of [...inviteesByUserId, hostId]) {`

Achieves the same result without needing to do the previous lines

navarrojandg commented on apps/api/src/modules/etl/events/create.ts:
```diff
@@ -84,7 +88,7 @@ export const etlEventsCreate = async (
         name,
         description,
-        guestsByUserId: [...inviteesByUserId],
+        guests: [...inviteesByUserId],
```
Same note here -- prefer `guestsByUserId` because it makes it really clear how we are representing the guests

navarrojandg commented on apps/api/src/modules/etl/event-plans/delete.ts:
```diff
@@ -0,0 +1,67 @@
+      const userEventPlanDocumentRef = firebaseFirestore.doc(
+        `/users/${uid}/event-plans/${eventPlanId}`
+      );
+      await firebaseFirestore.recursiveDelete(userEventPlanDocumentRef);
```
This should be its own try block. Doing so will make it clear which delete operation failed, did it fail because we couldn't delete the event-plan doc itself or because we couldn't delete the user event-plan doc.

try {
// delete event-plan doc
} catch(err) {
// error related to event-plan doc deletion

try {
// delete user event-plan doc
} catch(err) {
// error related to user event-plan doc deletion
}

navarrojandg commented on apps/api/src/modules/etl/events/delete.ts:
```diff
@@ -0,0 +1,65 @@
+      const userEventDocumentRef = firebaseFirestore.doc(
+        `/users/${uid}/events/${eventId}`
+      );
+      await firebaseFirestore.recursiveDelete(userEventDocumentRef);
```
Similar note as in event-plans/delete: should be its own try block

navarrojandg commented on apps/api/src/modules/etl/events/delete.ts:
```diff
@@ -0,0 +1,65 @@
+    if (guests !== undefined) {
+      guestsAndHost.push(...guests.map((value) => value.uid));
+    }
+    for (const uid of guestsAndHost) {
```
Similar to event-plans

`for (const uid of [...guestsByUserId, hostId]) {`

navarrojandg commented on apps/api/src/modules/etl/events/delete.ts:
```diff
@@ -0,0 +1,65 @@
+
+type Params = {
+  eventId: EventId;
+  guests: EventGuest[];
```
Similar note -- https://github.com/CSUN-COMP-490-Team-KBBQ/wya/pull/10/files#r846658919

navarrojandg reviewed (changes requested):
Overall don't think of the client as a source of truth, rely on the firestore (our db) as the source of truth. For example, don't rely on the client to provide us with a list of invitees or guests, rather just have the client provide the document id and we can fetch that information ourself from firestore.

ivancuadra869 commented:
> Overall don't think of the client as a source of truth, rely on the firestore (our db) as the source of truth. For example, don't rely on the client to provide us with a list of invitees or guests, rather just have the client provide the document id and we can fetch that information ourself from firestore.

Wow that's some great feedback man, thank you! I'm still learning throughout this whole process

ivancuadra869 committed:
Fix changes recommended by Jan

ivancuadra869 committed:
Add host check for deleting event/event-plans

ivancuadra869 committed:
Fix response code in api

For delete event/event-plan

jesc0bar commented:
> Finalized event was not deleted on my end.

Looks like the finalized event does get deleted but the event-plan for the same event does not.

jesc0bar commented on apps/api/src/interfaces/index.ts:
```diff
@@ -77,7 +77,7 @@ export type EventPlanDocument = {
   endDate: string;
   hostId: UserId;
-  invitees: UserId[] | Email[];
+  inviteesByUserId: UserId[] | Email[];
```
This could be confusing if we get emails and we're referring to it by userId.

navarrojandg replied:
Yes, don't want to store other people's emails on firestore, so we should let the api handle the conversion. For example when creating an event-plan we give a list of emails which our api converts to a list of user ids and stores in firestore. So our docs should store user ids exclusively when pertaining to other users

ivancuadra869 replied:
Yeah, I'm not sure why the email type is there, It was an over sight, I'll remove it

jesc0bar commented on apps/web/src/interfaces/index.ts:
```diff
@@ -77,7 +77,7 @@ export type EventPlanDocument = {
   endDate: string;
   hostId: UserId;
-  invitees: UserId[] | Email[];
+  inviteesByUserId: UserId[] | Email[];
```
same here

jesc0bar reviewed (commented):
looks good. functionality is still not entirely working on my end when testing - event-plan document does not get removed.

ivancuadra869 commented:
> > Finalized event was not deleted on my end.
> 
> Looks like the finalized event does get deleted but the event-plan for the same event does not.

Yeah, that's not wrong is it? If anything, creating an event should delete the event plan, right?

ivancuadra869 committed:
Remove Email[] type from inviteesByUserId

jesc0bar commented:
> > > Finalized event was not deleted on my end.
> > 
> > 
> > Looks like the finalized event does get deleted but the event-plan for the same event does not.
> 
> Yeah, that's not wrong is it? If anything, creating an event should delete the event plan, right?

That makes sense. I'll create a ticket for it. Thanks

navarrojandg committed:
Clean up event-plans

navarrojandg committed:
Clean up events

navarrojandg committed:
Remove event-plan when event is deleted

navarrojandg committed:
Add note for calling etlEventPlansDelete within etlEventsDelete

navarrojandg reviewed (approved):
One more smoke test and then feel free to merge onto staging !

ivancuadra869 committed:
Refactor EventDocument and add day create event

ivancuadra869 committed:
Fix event and eventPlan delete cloud functions

The recursive delete within the runTransaction was not safe, it kept timing out.
